### ISSUE TITLE
Find IPD setting

### ISSUE THREAD

**Issue by mungewell** [gusmanb/PSVRFramework#20]
According to:

**Comment by mungewell** [comment]
I should be clearer:
The Processing Unit renders left/right eye views and performs barrel distortion when running in Cinematic mode. If the PS4 setting 'IPD' affects the rendered view, then this is being communication to the PU via USB 'command'. We should be able to find this 'command' and duplicate it.

I don't have PS4, so if someone could try adjusting the IPD setting between max/min to see if it does really make a difference that would be helpful.

**Comment by gusmanb** [comment]
We would need an USB analyzer for this. If someone has a schematic or project link to create a cheap one that would be a very huge advance, I could then analyze all the packets through PS4/BBox.

**Comment by berlincount** [comment]
I have an http://openvizsla.org/ (see https://www.kickstarter.com/projects/bushing/openvizsla-open-source-usb-protocol-analyzer), if that would help?

**Comment by gusmanb** [comment]
@berlincount Do you have the ps vr and a ps4? if that's the case then YES!!! that could be extremely useful, if we can capture commands during the execution of concrete actions that would lead to a very large step forward on understanding all the USB protocol.

**Comment by berlincount** [comment]
I've got neither - yet. I'm waiting for a good bundle deal on the PS4 Pro after its release.

Do you live in Europe, or will you be on the Chaos Communcation Congress in Hamburg this year?

**Comment by gusmanb** [comment]
@berlincount Yes, I'm on Europe, I'm spanish, but no, I will not be at CCC congress.

**Comment by berlincount** [comment]
I could lend you my OpenVizsla, if you tread it nicely ;)

**Comment by gusmanb** [comment]
@berlincount That's a very generous offering. First let's try to advance without it (or may be I can get one in a near future) and if we get stucked at some point we can see to organize a shipment if you still don't have the PS VR (I prefer to "toy" with my equipment, if I break something that would be just my problem, I'm not telling I will break it, I work as hardware engineer too, but you know how these things are, the most stupid mistake can break an expensive piece of hardware).

Thanks a lot :)

**Comment by mungewell** [comment]
In command 0x21 '03: byte?, unknown, max value 50', with default value (from logs) is 32, 50-32 = 18... 

There appears to be some discrete settings that the PS4 actually uses, adjustable with +/-18 clicks.
https://www.reddit.com/r/PSVR/comments/58jb5y/ipdeyetoeye_measurement_number_cheat_sheet/

Tempted to set up a webcam looking into HMD to see if rendered image moves as this is changed.

How to measure your IPD at home
http://doc-ok.org/?p=898

**Comment by gusmanb** [comment]
Did a dirty test, just added a command to set it to 0/50 and worn the hdm. Changing it does nothing visible, looking at the edges of the lenses no pixel moves at all. Said that, I'm not sure if it's just a "placebo effect" but my eyes note a bit of tension when I set it to 0, so may be this is really IPD setting, if they did a perfect IPD correction as they should then they aren't just displacing the projection but distorting it so that will make sense, the center of the view is what has been displaced and the edges remain at the same place.

I already noticed this "tension" when I tested the cinematic report but discarded it as "just my mind playing tricks" as I saw nothing, but this starts to make a lot more sense...

If you have a cam which fits inside the hdm a pixel comparison test would be the perfect one, record it at 0 and at 50 and compare it pixel by pixel at the center of the projection, if something changes then it's the IPD. I only have cams on an iMac and it's mounted at the screen and also in a laptop and the same happens so I can't do the test until I go to an electroinc store to make an adapter for the PS4 camera.

**Comment by mungewell** [comment]
I have quite a few web cams, and this:
https://support.smarttech.com/en/hardware/other-hardware/document-cameras/document-camera-450

Which has a real good macro mode and optical zoom. Sounds like a project for tonight...

**Comment by gusmanb** [comment]
Yup! that's really nice, with that a perfectly steady image can be achieved 👍 

Going to sleep as it's 4AM here, will see the results tomorrow.

Good luck!

**Comment by mungewell** [comment]
I was able to focus quite close on the screen/one-eye, but was not able see any consistent movement of the image. This process was complicated by screen location is 'reset' whenever the 0x21 command is sent.

Just looking into headset and changing value, did not really seem to make a difference.

**Comment by mungewell** [comment]
So in my comment above I made some mistakes....

Firstly the unknown byte in question is Byte 0x03, which is set to 0x14 in the USB logs we have. It appears that it can change from 0 to 41 (0x29). 0x14 would be the half way setting.

@gusmanb suggested that byte 0x00 might be a bit field (we've seen 0x80 and 0xC0 set). Using the logic from LEDs then bit 0 would affect the first byte, and bit 7 the last. These might be the 8 bytes from 0x03 through 0x0B (ie ending with Mic sideband). 

Given that we've identified the 'worn bit' in 0xF0 report, the logs then show the transition from 0x80 to 0xC0, ie 'now' setting screen brightness - but who really knows what PS4 code it doing.

In which case we may well need to set extra bits in the bitfield to have the adjusted setting actually do something. I'll retry the IPD test with is set all on, and see if clearing to 0 prevents brightness and mic level being set).

**Comment by mungewell** [comment]
Getting back into PSVR 11 months later.... and still looking for the IPD control.

What I have discovered is that the lock feature only works when the Cinematic mode is in the state where the PU renders the gamepad icon if you turn too far from the 'screen'.

This mode is configured with:
`ep.write(b'\x11\x00\xaa\x08\x02\x00\x03\x11\x00\x00\x00\x00')`

Then lock can be set/cleared as follows. As noted above this value looks very like a bit field. Perhaps one of these bits also enables the IPD adjustment. With lock enabled we should be able to use a camera to view the OLEDs directly without any tremble/wobble from the acc/gyros.

Still think (/hope) that the 'mystery' byte does something...
```
		if options.lock:
			lock = 0x80
		else:
			lock = 0xc0

		if options.size:
			ep.write(b'\x21\x00\xaa\x10' + \
				bytearray([lock, options.size, options.dist, options.mist]) + \
				b'\x00\x00\x00\x00\x00\x00' + bytearray([options.bright]) + \
				b'\x00' + bytearray([options.hdmi]) + b'\x00\x00\x00')
```
